Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1356 @@
+  | exp exp                  { EApp (info $1 <-> info $2) $1 $2 }
+```
+
+But we'll immediately see that it will produce five shift/reduce conflicts.
+The ambiguity comes from the fact that `a b c` could be parsed either as `(a b) c` or `a (b c)`.
+If you analyze the output of `happy -i src/Parser.y`, you should see that it can either shift to another token or reduce it to a function application.
+
+Remember, shifting asks Happy to continue parsing, making the output right-associative, while reducing asks Happy to accept what we have so far, making the output left-associative.
+
+Since shifting is the default action, we will get the incorrect output with `let x = a b c`:
+
+```haskell
+Right
+  [ Dec _ (Name _ "x") [] Nothing (EApp _ (EVar _ (Name _ "a")) (EApp _ (EVar _ (Name _ "b")) (EVar _ (Name _ "c"))))
+  ]
+```
+
+Unfortunately, Happy doesn't have a `%reduce` directive, only a `%shift` directive.
+Furthermore, using precedences here will not help us.
+
+One feature supported by Happy is the `%prec` directive, which allows creating a placeholder precedence such as `%left APP`, and then using it like this:  `| exp exp %prec APP` (see [2.3.2](https://monlih.github.io/happy-docs/#_sec_precedences) in the Happy User Guide).
+It might cause shift/reduce conflicts later on, though, that could be fixed by listing all terminals in the precedence table.
+Sometimes it might be less annoying than refactoring the grammar, but it's not our case here.
+
+Instead of messing with precedences, we will make an observation that will allow us to refactor our grammar in an intelligent way to resolve this ambiguity: the right side of a function application will always be an "atom".
+
+To be more precise, ambiguities, in general, should appear every time we have a production such as `A B C ... exp` or `exp ... X Y Z`.
+Had we placed another token to the left or the right of `exp exp`, such as `exp exp in`, the ambiguity would be eliminated (in this specific case).
+
+The solution is to extract all "atoms" into one production, and use `atom` on the right side of the application:
+
+```happy
+exp :: { Exp L.Range }
+  : exp atom                 { EApp (info $1 <-> info $2) $1 $2 }
+  | atom                     { $1 }
+
+atom :: { Exp L.Range }
+  : integer                  { unTok $1 (\range (L.Integer int) -> EInt range int) }
+  | name                     { EVar (info $1) $1 }
+  | string                   { unTok $1 (\range (L.String string) -> EString range string) }
+  | '(' ')'                  { EUnit (L.rtRange $1 <-> L.rtRange $2) }
+  | '[' sepBy(exp, ',') ']'  { EList (L.rtRange $1 <-> L.rtRange $3) $2 }
+  | '(' exp ')'              { EPar (L.rtRange $1 <-> L.rtRange $3) $2 }
+```
+
+This will cause Happy to build a chain like `(((atom atom) atom) atom) atom` when parsing.
+Note that even though the left side was stated to be an `exp`, we must eventually reach an `atom` to terminate parsing.
+
+With this, function application should now be left-biased.
```

Comment:
Uh... perhaps s/left-biased/left-associative/, since we're talking about an operator, not a tree here.